### PR TITLE
Migrate to library javax.json for JSON objects

### PR DESCRIPTION
**Pull request by Beacze**

### CODE REVIEW

File: Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java [hunk 1]
```diff
@@ -564,11 +567,12 @@ public S findById(Integer id, boolean related) throws DataException {
     /**
      * Convert JSONObject object to bean object.
      *
+     * gi
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java]: What is the meaning of `gi` or is something going lost?

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 2]
```diff
@@ -707,39 +706,41 @@ public Process convertDtoToBean(ProcessDTO processDTO) throws DAOException {
...
         }
         return processDTO;
     }
 
-    private ProcessDTO convertRelatedJSONObjects(JSONObject jsonObject, ProcessDTO processDTO) throws DataException {
-        Integer project = getIntegerPropertyForDTO(jsonObject, "project.id");
-        processDTO.setProject(serviceManager.getProjectService().findById(project));
+    private ProcessDTO convertRelatedJSONObjects(JsonObject jsonObject, ProcessDTO processDTO) throws DataException {
+        Integer project = jsonObject.getInt("project.id");
+        if (project > 0) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: Should here not a check if value of property `project` is not `null` as comparing `null` with `0` can cause trouble?

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/index/type/AuthorityType.java [hunk 3]
```diff
@@ -11,22 +11,24 @@
...
 
-    @SuppressWarnings("unchecked")
     @Override
     public HttpEntity createDocument(Authority authority) {
 
-        JSONObject authorizationObject = new JSONObject();
-        authorizationObject.put("title", authority.getTitle());
-        authorizationObject.put("userGroups", addObjectRelation(authority.getUserGroups(), true));
+        JsonObject authorityObject = Json.createObjectBuilder()
+                .add("title", preventNull(authority.getTitle()))
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/index/type/AuthorityType.java]: Is formatting correct here and all other similar places? It is not indented to the property nor 4 space indent.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java (hunk 2 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]:
```
Integer projectId = process.getProject() != null ? process.getProject().getId() : 0;
```
Here it doesn't store anymore nulls - 0 instead.

**Comment by henning-gerhardt** [comment]
There are merge conflicts in

- Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/index/type/BaseType.java
- Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java
- Kitodo-DataManagement/src/test/java/org/kitodo/data/elasticsearch/index/type/BatchTypeTest.java
- Kitodo-DataManagement/src/test/java/org/kitodo/data/elasticsearch/index/type/ProjectTypeTest.java
- Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java
- Kitodo/src/main/java/org/kitodo/services/data/ProjectService.java
- Kitodo/src/main/java/org/kitodo/services/data/base/SearchService.java 

Please fix them.

**Comment by henning-gerhardt** [comment]
An additional merge conflict is in

- Kitodo/src/main/java/org/kitodo/services/data/UserService.java

Please fix / solve them too.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/index/type/BaseType.java [hunk 4]
```diff
@@ -109,8 +125,18 @@
      *            as Date
      * @return formatted date as String
      */
-    String formatDate(Date date) {
-        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
-        return dateFormat.format(date);
+    JsonValue getFormattedDate(Date date) {
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/index/type/BaseType.java]: As you changed the return value you must adjust the existing Javadoc too.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java [hunk 5]
```diff
@@ -156,26 +146,22 @@ public JSONObject findDocument(String query) throws DataException {
...
-            JSONObject jsonObject = (JSONObject) parser.parse(response);
-            if (jsonObject.containsKey("hits")) {
-                JSONObject hits = (JSONObject) jsonObject.get("hits");
-                JSONArray inHits = (JSONArray) hits.get("hits");
-                if (!inHits.isEmpty()) {
-                    return (JSONObject) inHits.get(0);
-                }
-            } else {
-                return jsonObject;
+        JsonReader jsonReader = Json.createReader(new StringReader(response));
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java]: Same here. `JsonReader` is closeable so using a try-with-resources statement for closing the object is maybe a good idea.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java [hunk 6]
```diff
@@ -220,26 +206,24 @@ public JSONObject findDocument(String query, String sort) throws DataException {
...
 
         String response = restClient.getDocument(query, sort, offset, size);
-        try {
-            JSONObject jsonObject = (JSONObject) parser.parse(response);
-            if (jsonObject.containsKey("hits")) {
-                JSONObject hits = (JSONObject) jsonObject.get("hits");
-                JSONArray inHits = (JSONArray) hits.get("hits");
-                if (!inHits.isEmpty()) {
-                    searchResults.addAll(inHits);
+        JsonReader jsonReader = Json.createReader(new StringReader(response));
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java]: Same here. `JsonReader` is closeable so using a try-with-resources statement for closing the object is maybe a good idea.

File: Kitodo-DataManagement/src/test/java/org/kitodo/data/elasticsearch/search/SearcherIT.java [hunk 7]
```diff
@@ -152,23 +154,24 @@ public void shouldFindDocumentByQuery() throws Exception {
...
+        jsonObject = result.getJsonObject("_source");
         id = getIdFromJSONObject(result);
         assertEquals("Incorrect result - id doesn't match to given plain text!", 1, id.intValue());
-        title = (String) jsonObject.get("title");
+        title = jsonObject.getString("title");
         assertEquals("Incorrect result - title doesn't match to given plain text!", "Batch1", title);
 
         query = "{\n\"match\" : {\n\"title\" : \"Nonexistent\"}\n}";
         result = searcher.findDocument(query);
+        System.out.println(result);
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/test/java/org/kitodo/data/elasticsearch/search/SearcherIT.java]: Is this `System.out.println()` call really needed? Maybe you should use a logger call?

File: Kitodo/src/main/java/de/sub/goobi/forms/IndexingForm.java [hunk 8]
```diff
@@ -406,14 +407,12 @@ private void resetGlobalProgress() {
...
-    private static String readMapping() throws ParseException {
-        JSONParser parser = new JSONParser();
+    private static String readMapping() {
         ClassLoader classloader = Thread.currentThread().getContextClassLoader();
         try (InputStream inputStream = classloader.getResourceAsStream("mapping.json")) {
             String mapping = IOUtils.toString(inputStream, "UTF-8");
-            Object object = parser.parse(mapping);
-            JSONObject jsonObject = (JSONObject) object;
-            return jsonObject.toJSONString();
+            JsonObject jsonObject = Json.createReader(new StringReader(mapping)).readObject();
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/IndexingForm.java]: Same here. `JsonReader` is closeable so using a try-with-resources statement for closing the object is maybe a good idea.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 9]
```diff
@@ -707,41 +706,41 @@ public Process convertDtoToBean(ProcessDTO processDTO) throws DAOException {
...
-        processDTO.setSortHelperMetadata(getIntegerPropertyForDTO(processJSONObject, "sortHelperMetadata"));
-        //TODO: remove access to database here, store path in index
-        processDTO.setTifDirectoryExists(false);
-        //processDTO.setTifDirectoryExists(
-        //checkIfTifDirectoryExists(processDTO.getId(), processDTO.getTitle(), processDTO.getProcessBaseUri()));
+        processDTO.setSortHelperArticles(processJSONObject.getInt("sortHelperArticles"));
+        processDTO.setSortHelperDocstructs(processJSONObject.getInt("sortHelperDocstructs"));
+        processDTO.setSortHelperImages(processJSONObject.getInt("sortHelperImages"));
+        processDTO.setSortHelperMetadata(processJSONObject.getInt("sortHelperMetadata"));
+        processDTO.setTifDirectoryExists(
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: In a previous commit you commeted this call out as it takes a long time and slowed down the displaying of list entries. Is this now solved?

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java [hunk 10]
```diff
@@ -65,16 +67,12 @@ public Long countDocuments() throws DataException {
...
 
         String response = restClient.countDocuments(overrideNullQuery(query));
         if (!response.equals("")) {
-            try {
-                JSONObject result = (JSONObject) parser.parse(response);
-                return (Long) result.get("count");
-            } catch (ParseException e) {
-                throw new DataException(e);
-            }
+            JsonReader jsonReader = Json.createReader(new StringReader(response));
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java]: `JsonReader` is a closeable class but the created object is never closed. Maybe you can use a try-with-resources statement?

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java [hunk 11]
```diff
@@ -120,19 +114,15 @@ private String overrideNullQuery(String query) {
...
-        JSONParser parser = new JSONParser();
 
         String response = restClient.getDocument(id);
         if (!response.equals("")) {
-            try {
-                return (JSONObject) parser.parse(response);
-            } catch (ParseException e) {
-                throw new DataException(e);
-            }
+            JsonReader jsonReader = Json.createReader(new StringReader(response));
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java]: Same here. `JsonReader` is closeable so using a try-with-resources statement for closing the object is maybe a good idea.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java [hunk 12]
```diff
@@ -89,20 +87,16 @@ public Long countDocuments(String query) throws DataException {
...
 
         String response = restClient.aggregateDocuments(overrideNullQuery(query), aggregation);
         if (!response.equals("")) {
-            try {
-                JSONObject result = (JSONObject) parser.parse(response);
-                return (JSONObject) result.get("aggregations");
-            } catch (ParseException e) {
-                throw new DataException(e);
-            }
+            JsonReader jsonReader = Json.createReader(new StringReader(response));
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/elasticsearch/search/Searcher.java]: Same here. `JsonReader` is closeable so using a try-with-resources statement for closing the object is maybe a good idea.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java (hunk 2 above)
Comment (henning-gerhardt) [reply Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: I hope, you are right as `Integer` can be `null`.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java (hunk 9 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: Yes, this one accesses database only if value is empty.

**Pull request merged**